[team7][BE 우디, 비비] issue-tracker PR (1)

bibi6666667 opened the pull request:
안녕하세요, 7팀의 issue-tracker 프로젝트 백엔드를 맡은 우디&비비 입니다😀
처음이자 마지막 PR인데, 커밋이 꼬여 이제야 PR 날리는 점 죄송합니다💦

<구현한 부분>
- AWS 
  - EC2 : VPC 분리 생성
  - S3 : 버킷을 활용한 이미지 업로드 기능
- Github OAuth & JWT
- Jasypt : 민감 정보 암호화
- JPA & QueryDSL로 DB CRUD와 필터링 로직 구현

잘못되거나 바람직하지 못한 부분 아낌없는 피드백 부탁드립니다!
감사합니다🙏

bibi6666667 committed:
feat : initial commit

bibi6666667 committed:
feat : 데이터베이스 연결 및 schema.sql 작성 #1

bibi6666667 committed:
feat : schema.sql의 일부 테이블 수정 #1

jihye-woo committed:
feat : ER diagram 구현 #1

bibi6666667 committed:
feat : ERD의 각 테이블을 domain 객체로 맵핑 #3

jihye-woo committed:
feat : response 래핑 클래스 ApiResponse 구현 #3

> ApiResponse 구현
  - 클라이언트에게 보낼 response를 래핑하기 위함
  - 모든 요청에 대한 response를 ApiResponse 객체에 담는 방식으로 통일
  - 이후 필요에 따라 status code 혹은 error message 필드 추가될 수 있음

jihye-woo committed:
feat : IssueController 구현 및 API 매핑 #3

> IssueController
  - Issue와 관련된 요청을 담당하는 컨트롤러
  - Issue에 대한 api와 매핑

> 내부 로직 미작성
  - 현재 기능 구현은 되어있지 않은 상태
  - service와 repository 계층과의 의존성 없음

jihye-woo committed:
feat : CommentController 구현 및 API 매핑 #3

> CommentController
  - Comment와 관련된 요청을 담당하는 컨트롤러
  - Commnet에 대한 api와 매핑

> CommentRequest 구현
  - comment 생성 요청에 대한 커맨드 객체

> 내부 로직 미작성
  - 현재 기능 구현은 되어있지 않은 상태
  - service와 repository 계층과의 의존성 없음

jihye-woo committed:
feat : MilestoneController 구현 및 API 매핑 #3

> MilestoneController
  - Milestone 관련된 요청을 담당하는 컨트롤러
  - Milestone에 대한 api와 매핑

> MilestoneRequest 구현
  - Milestone 생성 요청에 대한 커맨드 객체

> 내부 로직 미작성
  - 현재 기능 구현은 되어있지 않은 상태
  - service와 repository 계층과의 의존성 없음

jihye-woo committed:
feat : LabelController 구현 및 API 매핑 #3

> LabelController
  - Label 관련된 요청을 담당하는 컨트롤러
  - Label 대한 api와 매핑

> LabelRequest 구현
  - Label 생성 요청에 대한 커맨드 객체

> 내부 로직 미작성
  - 현재 기능 구현은 되어있지 않은 상태
  - service와 repository 계층과의 의존성 없음

jihye-woo committed:
feat : LabelController 구현 및 API 매핑 #3

> LoginController
  - Login 관련된 요청을 담당하는 컨트롤러
  - Login 대한 api와 매핑

> 내부 로직 미작성
  - 현재 기능 구현은 되어있지 않은 상태
  - service와 repository 계층과의 의존성 없음

jihye-woo committed:
feat : IssueRequest 커맨드 객체 구현 및 적용 #3

> IssueRequest 커맨드 객체
 - 이슈 생성에 필요한 필드 지정
 - 이슈 생성 요청에 대한 request를 커맨드 객체로 받기

bibi6666667 committed:
feat : OAuth 구현을 위한 클래스 추가 #4

bibi6666667 committed:
feat : Github OAuth 로 사용자 정보 받기 #4

jihye-woo committed:
feat : jasypt 암호화 설정

> jasypt 설정 추가 및 적용
  - github client secret key를 jasypt을 통해 암호화
  - jasypt password는 환경 변수 설정

jihye-woo committed:
feat : jasypt config에 pool size 설정 #4

> pool size 관련 에러 핸들링
  - "pool size must be set and > 0" 에러 발생
  - JasyptConfig에서 poolsize를 1로 고정

jihye-woo committed:
feat : oauth 관련 properties 파일 수정 #4

> github client id의 불필요한 문자열 제거

jihye-woo committed:
feat : logger 설정 및 properties 코드 정리 #4

> issuetracker 폴더 아래의 코드는 debug 로그로 설정

> application.properties 코드 정리

jihye-woo committed:
feat : GitHub에 종속적인 클래스 이름 수정 #4

> GitHub OAuth 로그인 관련 클래스명에 'GitHub' 추가
  - LoginController ➡ GitHubLoginController
  - LoginService ➡ GitHubLoginService

jihye-woo committed:
feat : jjwt 의존성 추가 및 Jwt 생성 로직 구현 #4

> jjwt 의존성
  - build.gradle에 jjwt 의존성 추가
  - application-jwt.properties로 jwt 관련 설정 추가

> JwtProvider 컴포넌트 클래스 구현
  - jwt 생성 로직 구현

jihye-woo committed:
feat : User 클래스 수정 #4

> GeneratedValue 어노테이션
  - id 값 자동 생성을 위해 추가

jihye-woo committed:
feat : GitHub OAuth 구현 #4

> GitHub OAuth 로직 구현
  - GitHubLoginController
  - GitHubLoginService

> GitHubLoginResponse 구현
  - GitHub 로그인 성공 시 반환하는 response의 wrapper 클래스

jihye-woo committed:
feat : GitHub OAuth 테스트 컨트롤러 구현 #4

> LoginController 구현
  - @Controller 어노테이션
  - login() : /login/github으로 리다이렉트

bibi6666667 committed:
feat : jwt토큰이 없을 때 발생할 Exception 생성 및 등록 #4

bibi6666667 committed:
feat : jwt토큰 검증에 필요한 클래스 생성  #4

bibi6666667 committed:
chore : jwt, response 패키지 생성 및 이동  #4

bibi6666667 committed:
feat : 로그인된 유저에게 jwt토큰 발행 #4

bibi6666667 committed:
feat : Mock API 배포를 위한 스크립트 작성 #3

bibi6666667 committed:
chore : GithubLoginResponse -> GitHubUserResponse로 이름 변경  #3

bibi6666667 committed:
chore : import한 Strings 클래스 변경 #3

bibi6666667 committed:
feat : API 응답을 위한 Response클래스 생성 #3

bibi6666667 committed:
chore : 필드의 오타 수정 #3

bibi6666667 committed:
feat : MockAPI사용을 위해 GET 요청의 반환값 적용 #3

bibi6666667 committed:
feat : MockAPI사용을 위해 GET 요청의 반환값 적용 by Woody #3

jihye-woo committed:
feat : Issue 목록 조회 기능 추가 #4

> IssueService 구현
  - IssueService::getOpenedIssues
    - open 된 이슈 목록 조회
  - IssueService::getClosedIssues
    - close 된 이슈 목록 조회

> IssueRepository 구현
  - derived query method 작성

bibi6666667 committed:
feat : API 테스트를 위한 테스트 데이터 삽입 #3

bibi6666667 committed:
feat : @Id import 변경 및 @GeneratedValue 추가 #3

bibi6666667 committed:
refactor : User의 userId -> loginId로 필드명 변경 #3

bibi6666667 committed:
refactor : User의 userId -> loginId로 필드명 변경 #3

bibi6666667 committed:
feat : Issue클래스에 기본생성자, getter, toString() 생성 #3

bibi6666667 committed:
chore : 코드 재정렬 #3

bibi6666667 committed:
feat : Issue CRUD를 위한 메서드 수정 #3

bibi6666667 committed:
feat : ApiResponse가 매개변수 없이도 반환될 수 있도록 변경 #3

bibi6666667 committed:
feat : userId의 타입을 String->Long으로 변경 #3

jihye-woo committed:
feat : jjwt 의존성 변경 #7

> jjwt 의존성 추가
  - jjwt-impl 추가
  - jjwt-jackson 추가

> jjwt 버전 변경
  - 상위 버전인 11.0로 변경

jihye-woo committed:
feat : jwt에 사용자 정보 추가 #7

> jwt 생성 시 user 정보 추가
  - "user" 커스텀 claim을 만들어 유저 정보 추가

> jwt 유효성 검사 후 user 정보 추출
  - HttpServletRequest의 속성에 user 정보 추가
  - 컨트롤러 레이어에서 @RequestAttribute 어노테이션을 통해 user 정보 추출

jihye-woo committed:
feat : comment 관련 url, 인터셉터 등록 #7

> JwtConfig에서 인터셉터 url 추가 등록
  - /comment/**

jihye-woo committed:
chore : lombok 의존성 추가

bibi6666667 committed:
feat : UserResponse의 필드명 userId -> loginId로 변경 #3

bibi6666667 committed:
feat : 찾는 이슈가 없을 때 발생시킬 예외 추가 #3

jihye-woo committed:
refactor : lombok 적용

> 기존 코드에서 lombok으로 대체할 수 있는 부분는 적용
 - Comment과 연관된 기능에 한정하여 적용

bibi6666667 committed:
feat : 엔티티 등록을 위한 기본생성자와 어노테이션 추가 #3

bibi6666667 committed:
feat : Repository 등록을 위한 어노테이션 추가 및 클래스 생성 #3

jihye-woo committed:
feat : CommentRequest 유저 관련 필드 수정 #7

> userId 필드 삭제
  - JWT를 통해서 유저 정보를 받아오기 때문에 불필요해짐

> createdAt 값
  - 유저에게서 받아오지 않고 생성자 내부에서 초기화

bibi6666667 committed:
feat : 각 객체를 Response객체로 변환하는 메서드 추가 #3

bibi6666667 committed:
feat : Issue에 User 1:1맵핑 및 getter, setter 추가 #3

bibi6666667 committed:
feat : IssueLabel 클래스 추가 #3

bibi6666667 committed:
feat : IssueRequest에 기본생성자, toString() 추가 #3

bibi6666667 committed:
chore : 불필요한 import 삭제 #3

bibi6666667 committed:
feat : IssueService의 get, delete 메서드 기능구현 #3

jihye-woo committed:
feat : User 클래스 필드명 수정 #7

> user_id ➡️ login_id 필드명 변경
 - id와 혼동이 될 것을 고려
 - 변경된 user 테이블 컬럼명에 따라 관련된 User 클래스 필드명도 함께 수정

bibi6666667 committed:
feat : Label의 IssueLabel에 대한 1:N맵핑 추가 #3

bibi6666667 committed:
feat : Issue update 메소드 연결 #3

bibi6666667 committed:
feat : Issue update - title 구현 #3

bibi6666667 committed:
feat : LabelService 클래스 추가 #3

bibi6666667 committed:
feat : Issue 삭제 시 관련 comment, assignee, label도 삭제되도록 CASCADE옵션 부여 #3

jihye-woo committed:
feat : Comment 서비스와 레포지토리 클래스 구현 #7

> Comment 기능 구현을 위해서 서비스, 레포지토리 레이어 생성
  - Comment 생성 기능 구현 완료
  - Comment 조회 기능 미완성
    ➡️ DB에서 조회한 comment 데이터를 클라이언트에서 원하는대로 dto에 래핑해서 보내는 작업 필요

jihye-woo committed:
chore : 커스텀 배너 적용

> 안 예쁜 spring 기본 배너 대신, "WoodyBibi" 문자열로 커스텀 배너 적용

jihye-woo committed:
feat : Comment 수정하기 기능 구현 #7

> EditedComment 커맨드 객체 생성
  - 수정할 comment에 대한 정보를 담는 클래스 생성

> Comment 수정에 성공하면 HTTP code 204 반환(no content)

> CommentService에 Transaction 처리
  - 생성과 수정 작업은 Transactional

jihye-woo committed:
feat : Comment 생성에 대한 HTTP 상태 코드 리턴 #7

> Comment를 성공적으로 생성할 경우 201 상태코드 리턴

jihye-woo committed:
feat : Comment 삭제 기능 구현 #7

> CommentController::deleteComment 구현
  - 상태코드 204 (no content) 반환

jihye-woo committed:
feat : CommentReponse 구현 #7

> lombok 적용
  - @getter
  - @AllArgsConstructor

jihye-woo committed:
feat : Label 생성하기 기능 구현 #8

> Label 생성 기능 구현
  - LabelController::createLabel
  - LabelService::create
  - 서비스 레이어에서 생성 작업에 대한 transactional 처리

> Label::create
  - static 메소드로 변경

> Label 생성에 성공하면 HTTP code 204 반환(no content)

jihye-woo committed:
feat : Label 조회하기 기능 구현 #8

> Label 조회 기능 구현
  - LabelController::getLabels

jihye-woo committed:
feat : Label 수정하기 기능 구현 #8

> Label 수정 기능 구현
  - LabelController::editLabel

> 클라이언트에서 label의 id 값을 가지고 있어야 함
  - label 생성 시, id를 포함한 label 정보를 리턴

jihye-woo committed:
feat : Milestone 생성하기 기능 구현 #9

> Milestone 생성 기능 구현
  - MilestoneController::createMilestone
  - MilestoneService::create

> MilestoneRequest::create 추가
  - Milestone 객체를 생성하는 정적 팩토리 메소드 추가

bibi6666667 committed:
feat : Issue에 Assignee 리스트 추가를 위한 클래스 추가 #3

bibi6666667 committed:
feat : Issue의 Assignee 등 수정 기능 구현 #3

bibi6666667 committed:
feat : assignee 테스트 데이터 추가 #3

bibi6666667 committed:
chore : 미사용 메서드 삭제 #3

jihye-woo committed:
feat : json 속성 설정 #9

> Json 설정 관련 어노테이션 추가
  - JsonProperty
  - JsonFormat

jihye-woo committed:
chore : validation 의존성 추가 #9

> build gradle에 validation 의존성 추가

jihye-woo committed:
feat: 페이징 util 클래스 구현 #9

> PaginationUtil 클래스 구현
  - descendingPageable()
    : 내림차순 Pageable 객체를 리턴하는 메소드 구현

jihye-woo committed:
feat: Milestone 조회하기 기능 구현 #9

> Milestone 전체 조회 기능 구현
  - MilestoneController::getMilestones
  - MilestoneService::getMilestones
  - 페이징 기능 구현, 현재 디폴트 페이지를 0으로 하드코딩

> Milestone 단일 조회 기능 구현
  - MilestoneController::getMilestone
  - MilestoneService::getMilestone

jihye-woo committed:
feat : Milestone 수정하기 기능 구현 #9

> EditedMilestone 커맨드 객체 구현
  - milestone 수정 정보를 담는 커맨드 객체

> Milestone 수정 기능 구현
  - MilestoneController::editMilestone
  - MilestoneService::edit

jihye-woo committed:
feat : Milestone 생성자 추가 #9

> Milestone 생성자 추가
  - id, openIssue, closeIssue 값을 제외한 파라미터 변수

jihye-woo committed:
feat : Milestone 삭제하기 기능 구현 #9

> Milestone 삭제 기능 구현
  - MilestoneController::deleteMilestone
  - MilestoneService::delete

bibi6666667 committed:
feat : IssueLabel 테스트용 데이터 삽입 #3

bibi6666667 committed:
feat : Issue의 Label 수정 기능 구현 #3

bibi6666667 committed:
feat : Github OAuth Client 정보들을 환경변수로 치환

jihye-woo committed:
feat : UserResponse 정적 팩토리 메소드 생성 #3

> UserResponse::create 팩토리 메소드 생성

jihye-woo committed:
feat : LabelResponse에 id 추가 #8

bibi6666667 committed:
feat : API - Issue의 Label 수정 기능 구현 완료 #3

bibi6666667 committed:
feat : Issue의 Label 수정을 위한 테스트 데이터 추가 #3

bibi6666667 committed:
feat : GitHub iOS 로그인을 위한 메서드 구현 #4

jihye-woo committed:
refactor : access token 발급 메소드 중복 제거 #4

> GitHubLoginService::accessToken
  - accessTokenForWeb(), accessTokenForIOS() 중복 로직 통합

bibi6666667 committed:
feat : 요청을 받는 AssigneeRequest 클래스 생성 #4

bibi6666667 committed:
feat : Issue에 Comment List 맵핑 후 조회 기능 구현 #4

bibi6666667 committed:
feat : comment 테스트용 데이터 추가 #4

jihye-woo committed:
feat : 이슈 작성자 전체 목록 조회 #14

> 현재 오픈된 이슈들의 작성자 목록 조회
  - /user/author/open
  - UserController::getAllOpenIssueAuthor

> 현재 클로즈된 이슈들의 작성자 목록 조회
  - /user/author/close
  - UserController::getAllCloseIssueAuthor

jihye-woo committed:
feat : Cors 설정 추가

> CorsConfig 클래스 구현
  - Cors origin 지정

jihye-woo committed:
feat : jjwt-jackson 의존성 제거 #4

> JacksonSerializer, JacksonDeserializer 커스텀 클래스 구현 및 치환

jihye-woo committed:
feat : UserResponse에 id 추가 #3

> 유저 정보 전달 시, 사용자의 id 데이터까지 전달

bibi6666667 committed:
chore : jjwt dependency 0.11.2 -> 0.10.8

bibi6666667 committed:
refactor : jjwt dependency 에러 해결을 위한 메서드 직접 구현

jihye-woo committed:
fix : resolve JSON Deserialize issue!!

jihye-woo committed:
feat : Comment 생성 시 Comment 객체 리턴 #7

> CommentController::createComment
  - 리턴 타입 : ApiResponse

> CommentService::create
  - 리턴 타입 : Comment

jihye-woo committed:
feat : DB save 작업 후 반환 값 변경 #3

> Comment 데이터 생성 및 DB 저장 후 반환 타입
  - CommentResponse

> Label 데이터 생성 및 DB 저장 후 반환 타입
  - LabelResponse

> Milestone 데이터 생성 및 DB 저장 후 반환 타입
  - MilestoneResponse

bibi6666667 committed:
feat : iOS 로그인을 위한 메서드 추가 #4

jihye-woo committed:
feat : milestone 수정과 삭제 시, 트랜젝션 처리 #9

> @Transactional 추가
 - MilestoneService::edit
 - MilestoneService::delete

bibi6666667 committed:
feat : Issue 생성 메서드 수정 #3

bibi6666667 committed:
refactor : 생성자, getter, setter에 롬복 사용

bibi6666667 committed:
chore : 댓글 생성일자 JSON 반환 형식 지정 #3

bibi6666667 committed:
feat : Issue의 Assignee 수정 메서드 로직 변경 #3

jihye-woo committed:
chore : aws 설정 파일 추가 #15

> application-aws.properties 생성
  - S3에 대한 모든 권한을 가진 IAM 유저의 access key와 secret key 저장
  - 사용할 s3 버킷에 대한 region, bucket 정보 저장

jihye-woo committed:
chore : aws 의존성 추가 #15

> aws-java-sdk-s3 의존성 추가
  - BOM을 이용해 자동으로 프로젝트 sdk 의존성에 맞는 버전을 찾아줌

jihye-woo committed:
feat : S3Client 빈 생성 #15

> S3Client 클래스 구현
   - AmazonS3을 사용한 S3 이미지 업로드 로직 구현
   - 업로드가 끝나면 생성된 image url을 반환

> AwsConfig에서 S3Client 빈 생성
   - application-aws.properties의 aws 정보들을 통해
    AmazonS3 객체를 생성하고 S3Client 객체에 주입

jihye-woo committed:
feat : 이미지 업로드 관련 커스텀 예외 생성 #15

> ImageUploadException 커스텀 예외 구현

> GlobalExceptionHandler에서 예외 처리

jihye-woo committed:
feat : S3 파일 업로드 요청 처리 #15

> FileController::upload
  - 파일 업로드 POST 요청 처리

> S3FileService::upload
  - 파일 업로드 로직 구현
  - uploadFileToS3() 메소드에서 비동기로 S3에 이미지 업로드

bibi6666667 committed:
chore : Domain 중심으로 프로젝트 패키지 구조 개편

jihye-woo committed:
temp : 웹 테스트를 위해 Jwt 유효성 검사를 일시적으로 중단

> 프론트에서 oauth 구현이 완료되면 다시 유효성 검사 진행할 예정

jihye-woo committed:
feat : 작성자, 담당자 전체 조회 기능 구현 #3

> UserController::getAllAssignee
  - 담당자 전체 조회(open + close)

> UserController::getAllAuthor
  - 작성자 전체 조회(open + close)

bibi6666667 committed:
feat : 라벨 필터링 환경설정 및 메서드 추가 #14

bibi6666667 committed:
feat : 라벨 필터링 테스트 설정 및 메서드 추가 #14

bibi6666667 committed:
delete : LabelRepositorySupportTest.java 삭제 #14

bibi6666667 committed:
feat : 필터링 기능 구현을 위해 IssueRepositorySupport 생성 #14

bibi6666667 committed:
feat : Issue 필터링 기능 구현 #14

bibi6666667 committed:
feat : Issue 필터링 기능 구현을 위한 테스트 데이터 추가 #14

jihye-woo committed:
refactor : 모든 Issue 필터링 기능에 querydsl 적용 #14

> IssueRepositorySupport 메소드 추가 구현
  - querydsl를 사용하여 쿼리문 작성
  - findByAssignee() : 담당자를 기준으로 필터링
  - findByCommentUserId() : 코멘트 작성자를 기준으로 필터링

> IssueService 메소드 수정
  - 모든 필터링 기능은 IssueRepositorySupport 메소드 사용하도록 변경

jihye-woo committed:
refactor : Issue ➡️ IssueResponse 래핑 코드 수정 #14

> List<Issue> ➡️ List<IssueResponse>
  - stream 적용
  - convertToResponse 함수 생성해서 공통 코드 분리

jihye-woo committed:
feat : 이슈 필터 클래스 IssueFilter 구현 #14

> 필터링 항목을 필드 변수로 가짐

jihye-woo committed:
feat : multiple filtering issue #14

> IssueRepositorySupport::getFilteredIssues
  - 모든 이슈 요소들을 한꺼번에 필터링 처리

> PredicateBuilder 클래스 구현
  - BooleanExpression 타입 where 절을 생성

jihye-woo committed:
feat : 이슈 멀티 필터링 기능 구현 #14

> IssueController::getFilteredIssues
  - url : /issue/filter
  - 필터링 필드를 쿼리 스트링으로 전달 받음
  - 쿼리 스트링들은 IssueFilter 객체와 매핑

> IssueService::getFilteredIssues
  - issueRepositorySupport::getFilteredIssues 메소드 적용하여 필터링 결과값 리턴

jihye-woo committed:
feat : 수정, 삭제 요청 시 사용자 권한 확인 #3

> NotAuthorizedUserException 구현
  - 사용자 권한 관련 예외

> Issue, Comment 수정 삭제 요청 시 사용자 권한 확인
  - Service layer에서 사용자 권한 검사
  - 요청에 대한 권한이 없을 경우 NotAuthorizedUserException 발생

ksundong commented:
정식 리뷰기간은 끝났지만 고생많으셨습니다.
시간 나는 대로 짬짬이 리뷰하도록 할게요~

jihye-woo committed:
fix : fix typo in property key name #18

> github.client.id.secrets.ios 에서 github.client.secrets.ios로 수정

jihye-woo committed:
fix : 이슈 조회시 발생하는 NPE 해결 #18

> issueRequest 리스트 타입 필드 초기화
  - 일시적으로 NPE를 해결하기 위한 방법
  - 이후 조회 로직 전체를 변경할 예정

jihye-woo committed:
fix : activate BearAuthInterceptor #18

> auth 로그인 후 사용할 수 있는 서비스에 대해서 Interceptor 적용

> Interceptor에서 실제 사용자 정보 추출
  - 기존에 디폴트 값으로 하드코딩 되어있던 사용자 정보 대신 jwt 토큰에서 직접 사용자 정보를 추출하도록 변경

ksundong commented on .idea/workspace.xml:
```diff
@@ -0,0 +1,96 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
이 파일이 버전 관리 되어야 하는지 충분히 의논이 된 상태일까요?

ksundong commented on BE/build.gradle:
```diff
@@ -0,0 +1,72 @@
+plugins {
+    id 'org.springframework.boot' version '2.5.0'
```
되도록이면 0버전은 비추합니다.
보통 문제가 있는 경우가 많거든요.

ksundong commented on BE/build.gradle:
```diff
@@ -0,0 +1,72 @@
+    id 'org.springframework.boot' version '2.5.0'
+    id 'io.spring.dependency-management' version '1.0.11.RELEASE'
+    id 'java'
+    id "com.ewerk.gradle.plugins.querydsl" version "1.0.10"
```
해당 플러그인 적용없이 QueryDSL을 적용할 수 있으니 찾아서 해결해보시길 바랍니다.
약 2~3년 전에 해결된 이슈고, 그로 인해서 해당 플러그인은 버전 업그레이드가 전혀 되고 있지 않습니다.

ksundong commented on BE/build.gradle:
```diff
@@ -0,0 +1,72 @@
+    runtimeOnly 'io.jsonwebtoken:jjwt-impl:0.10.8'
+
+    // jasypt
+    implementation 'com.github.ulisesbocchio:jasypt-spring-boot-starter:3.0.0'
```
오 jasypt군요.

ksundong commented on BE/src/main/java/com/codesquad/issuetracker/auth/AuthorizationExtractor.java:
```diff
@@ -0,0 +1,25 @@
+@Component
+public class AuthorizationExtractor {
+    public static final String AUTHORIZATION = "Authorization";
+    public static final String ACCESS_TOKEN_TYPE = AuthorizationExtractor.class.getSimpleName();
```
`ACCESS_TOKEN_TYPE`의 의미하는 것이랑 대입된 변수랑 매칭이 잘 안되는데 설명좀 해주세요~

ksundong commented on BE/src/main/java/com/codesquad/issuetracker/auth/AuthorizationExtractor.java:
```diff
@@ -0,0 +1,25 @@
+        Enumeration<String> headers = request.getHeaders(AUTHORIZATION);
+        while (headers.hasMoreElements()) {
+            String value = headers.nextElement();
+            if (value.toLowerCase().startsWith(type.toLowerCase())) {
```
시작만 하면 되는건가요?

ksundong commented on BE/src/main/java/com/codesquad/issuetracker/auth/AuthorizationExtractor.java:
```diff
@@ -0,0 +1,25 @@
+                return value.substring(type.length()).trim();
+            }
+        }
+        return Strings.EMPTY;
```
해당 클래스가 어디서 오는지 꼭 확인해보시기 바랍니다.
적절한 것이 없다면 직접 정의해도 되고, 스프링에서 제공해주는 경우도 있는지 찾아보는 것도 좋습니다.

ksundong commented on BE/src/main/java/com/codesquad/issuetracker/auth/BearerAuthInterceptor.java:
```diff
@@ -0,0 +1,54 @@
+
+    @Override
+    public boolean preHandle(HttpServletRequest request, HttpServletResponse response, Object handler) {
+
```
의미 없는 빈 라인을 두는 것은 가급적이면 지양해주세요~

ksundong commented on BE/src/main/java/com/codesquad/issuetracker/auth/BearerAuthInterceptor.java:
```diff
@@ -0,0 +1,54 @@
+    @Override
+    public boolean preHandle(HttpServletRequest request, HttpServletResponse response, Object handler) {
+
+        request.setAttribute("user", getDefaultUser());
```
보통 어딘가의 key로 쓰이는 값은 매칭되어 사용되는 구간이 존재하기 마련입니다.
그래서 저는 상수로 꺼내서 쓰는걸 선호하는 편이에요.

ksundong commented on BE/src/main/java/com/codesquad/issuetracker/auth/BearerAuthInterceptor.java:
```diff
@@ -0,0 +1,54 @@
+        log.debug("User info from JWT : {}", user);
+
+        return user;
+    }
```
해당 메서드들이 사용되고 있나요?

ksundong commented on BE/src/main/java/com/codesquad/issuetracker/auth/JwtProvider.java:
```diff
@@ -0,0 +1,73 @@
+    private final long validityInMilliseconds;
+    private final String secretKey;
+
+    public JwtProvider(@Value("&{security.jwt.token.secret-key}") String secretKey,
```
```suggestion
    public JwtProvider(@Value("${security.jwt.token.secret-key}") String secretKey,
```

오타아닌가요?

ksundong commented on BE/src/main/java/com/codesquad/issuetracker/auth/JwtProvider.java:
```diff
@@ -0,0 +1,73 @@
+        Date now = new Date();
+        Date validity = new Date(now.getTime() + validityInMilliseconds); // 유효시간 (지금 + 유효기간)
+        log.info("now: {}", now);
+        log.info("validity: {}", validity);
```
info 레벨로 로깅할 필요는 없지 않을까요?

ksundong commented on BE/src/main/java/com/codesquad/issuetracker/auth/controller/GitHubLoginController.java:
```diff
@@ -0,0 +1,39 @@
+        logger.debug("iOS code : {} ", code);
+        GitHubUserResponse response = loginService.loginIOS(code);
+        return ApiResponse.ok(response);
+    }
```
왜 두 개를 따로 구분했는지 궁금합니다. ㅎㅎ
redirect 경로가 달라서일까요?

ksundong commented on BE/src/main/java/com/codesquad/issuetracker/auth/controller/GitHubLoginController.java:
```diff
@@ -0,0 +1,39 @@
+
+
+    @GetMapping("/github")
+    public ApiResponse<GitHubUserResponse> githubLogin(@RequestParam String code) {
```
이 `ApiResponse` 라는 타입은 왜 사용하시는지 잘 모르겠습니다.
제 기준에서 레거시 느낌이 강하게 나는 패턴입니다.
어디서 이것을 사용하는 것을 보셨는지 잘 모르겠지만, 이것을 사용했을 때와 사용하지 않았을 때를 구분해서 생각해보시면 좋을 것 같습니다.

ksundong commented on BE/src/main/java/com/codesquad/issuetracker/auth/controller/LoginController.java:
```diff
@@ -0,0 +1,31 @@
+    public LoginController(Environment environment) {
+        this.CLIENT_ID = environment.getProperty("github.client.id");
+        this.CLIENT_ID_IOS = environment.getProperty("github.client.id.ios");
+    }
```
this는 빼줘도 됩니다. 그리고 this 참조에 대해서도 공부해보는게 좋을 것 같네요.
Environment를 직접 주입받아 할당받는 방식보다는, 애노테이션을 활용하는 방법을 추천드리겠습니다.
```suggestion
    public LoginController(@Value("github.client.id") String clientId, @Value("github.client.id.ios") String iOSClientId) {
        this.CLIENT_ID = environment.getProperty("github.client.id");
        this.CLIENT_ID_IOS = environment.getProperty("github.client.id.ios");
    }
```

그리고 property의 계층구조도 생각해보시면 좋을 것 같습니다.
`github.client.id` → `github.client.id.web` 이 더 맞을 것 같습니다.

ksundong commented on BE/src/main/java/com/codesquad/issuetracker/auth/controller/LoginController.java:
```diff
@@ -0,0 +1,31 @@
+    @GetMapping("/iOS")
+    public String loginIOS() {
+        return "redirect:" + GITHUB_URL + "?client_id=" + CLIENT_ID_IOS;
+    }
```
중복 코드가 발생하고 있는데, 어떻게 처리하면 좋을까요?

ksundong commented on BE/src/main/java/com/codesquad/issuetracker/auth/domain/GitHubUser.java:
```diff
@@ -0,0 +1,34 @@
+
+public class GitHubUser {
+
+    private String login;
```
login이라는 것이 의미하는 바가 잘 와닿지 않습니다.
좀 더 구체적으로 네이밍을 하는게 좋을 것 같아요.

ksundong commented on BE/src/main/java/com/codesquad/issuetracker/auth/exception/TokenEmptyException.java:
```diff
@@ -0,0 +1,9 @@
+    private static final String TOKEN_EMPTY = "토큰이 존재하지 않습니다";
+    public TokenEmptyException() {
+        super(TOKEN_EMPTY);
+    }
```
```suggestion
    private static final String TOKEN_EMPTY = "토큰이 존재하지 않습니다";
    
    public TokenEmptyException() {
        super(TOKEN_EMPTY);
    }
```

ksundong commented on BE/src/main/java/com/codesquad/issuetracker/auth/request/AccessTokenRequest.java:
```diff
@@ -0,0 +1,36 @@
+    private final String client_id;
+
+    @JsonProperty("client_secret")
+    private final String client_secret;
```
네이밍에 있어서 신경 써주시면 좋을 것 같습니다.
```suggestion
    @JsonProperty("client_id")
    private final String clientId;

    @JsonProperty("client_secret")
    private final String clientSecret;
```

ksundong commented on BE/src/main/java/com/codesquad/issuetracker/auth/response/AccessTokenResponse.java:
```diff
@@ -0,0 +1,51 @@
+
+    public AccessTokenResponse() {
+
+    }
```
```suggestion
    public AccessTokenResponse() {}
```

사용하지 않는 빈 라인은 제거해주어도 좋고, 블럭 내부가 비어있는 경우에는 한 줄로 써줘도 무방합니다.

ksundong commented on BE/src/main/java/com/codesquad/issuetracker/auth/response/AccessTokenResponse.java:
```diff
@@ -0,0 +1,51 @@
+//        this.accessToken = accessToken;
+//        this.scope = scope;
+//        this.tokenType = tokenType;
+//    }
```
사용하지 않는 코드라면 바로바로 제거해주세요.

ksundong commented on BE/src/main/java/com/codesquad/issuetracker/auth/response/AccessTokenResponse.java:
```diff
@@ -0,0 +1,51 @@
+    @JsonSetter("token_type")
+    public void setTokenType(String tokenType) {
+        this.tokenType = tokenType;
+    }
```
아마 JsonProperty와 JsonSetter 둘 중에 하나만 사용해주면 될텐데 아니었나요?

ksundong commented on BE/src/main/java/com/codesquad/issuetracker/auth/response/GitHubUserResponse.java:
```diff
@@ -0,0 +1,32 @@
+public class GitHubUserResponse {
+
+    private String jwt;
+    private User user;
```
User를 그대로 리턴해주는데, 왠만하면 Response에는 Domain 혹은 Entity가 그대로 리턴되는 구조는 바람직하지 않습니다.

ksundong commented on BE/src/main/java/com/codesquad/issuetracker/auth/response/GitHubUserResponse.java:
```diff
@@ -0,0 +1,32 @@
+
+    public static GitHubUserResponse create(String jwt, User user, String type) {
+        return new GitHubUserResponse(jwt, user, type);
+    }
```
자바 컨벤션상 static 메서드는 일반 메서드보다 위에 위치해야 합니다.

ksundong commented on BE/src/main/java/com/codesquad/issuetracker/auth/serialize/JacksonDeserializer.java:
```diff
@@ -0,0 +1,77 @@
+import java.util.Collections;
+import java.util.Map;
+
+public class JacksonDeserializer<T> implements Deserializer<T> {
```
이 클래스를 직접 구현해주신 이유가 무엇일까요?

ksundong commented on BE/src/main/java/com/codesquad/issuetracker/auth/service/GitHubLoginService.java:
```diff
@@ -0,0 +1,101 @@
+    private final String GITHUB_CLIENT_SECRETS;
+
+    private final String GITHUB_CLIENT_ID_IOS;
+    private final String GITHUB_CLIENT_SECRETS_IOS;
```
```suggestion
    private static final String GITHUB_ACCESS_TOKEN_URL = "https://github.com/login/oauth/access_token";
    private static final String GITHUB_USER_URL = "https://api.github.com/user";
    
    private final UserRepository userRepository;
    private final JwtProvider jwtProvider;
    private final String githubClientId;
    private final String githubClientSecret;
    private final String iOSGithubClientId;
    private final String iOSGithubClientSecret;
```

Screaming snake case는 상수에만 사용해주세요.

ksundong commented on BE/src/main/java/com/codesquad/issuetracker/auth/service/GitHubLoginService.java:
```diff
@@ -0,0 +1,101 @@
+        User user = signUp(gitHubUser).orElseThrow(IllegalArgumentException::new);
+        String jwt = jwtProvider.createJwt(user);
+        return GitHubUserResponse.create(jwt, user, type);
+    }
```
private 메서드가 너무 많습니다.
각각을 다른 클래스로 꺼내야 하는건 아닌지 생각해봐주세요.

ksundong commented on BE/src/main/java/com/codesquad/issuetracker/config/AwsConfig.java:
```diff
@@ -0,0 +1,40 @@
+    private String region;
+
+    @Value("${cloud.aws.s3.bucket}")
+    private String bucket;
```
이것도 생성자 인젝션으로 풀어낼 수 있습니다.

ksundong commented on BE/src/main/java/com/codesquad/issuetracker/config/CorsConfig.java:
```diff
@@ -0,0 +1,16 @@
+    @Override
+    public void addCorsMappings(CorsRegistry registry) {
+        registry.addMapping("/**")
+                .allowedOrigins("http://localhost:3000", "https://3.35.137.242");
```
설정 파일로 추출하는 것은 어떨까요?

ksundong commented on BE/src/main/java/com/codesquad/issuetracker/config/JasyptConfig.java:
```diff
@@ -0,0 +1,29 @@
+import org.springframework.context.annotation.Configuration;
+
+@Configuration
+@EncryptablePropertySource(name = "EncryptedProperties", value = "classpath:application-jasypt.properties")
```
application prefix는 환경별 설정을 할 때 주로 이용됩니다~

ksundong replied:
이렇게 로딩해서 쓸 것이라면, 붙여주지 않아도 괜찮아요.

ksundong commented on BE/src/main/java/com/codesquad/issuetracker/config/JasyptConfig.java:
```diff
@@ -0,0 +1,29 @@
+    private String encryptKey;
+
+    @Bean("jasyptStringEncryptor")
+    public StringEncryptor stringEncryptor(){
```
```suggestion
    public StringEncryptor stringEncryptor() {
```

ksundong commented on BE/src/main/java/com/codesquad/issuetracker/domain/assignee/Assignee.java:
```diff
@@ -0,0 +1,51 @@
+@Entity
+public class Assignee {
+
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+    @JoinColumn(name = "issue_id")
+    private Long issueId;
+    @JsonProperty("user_id")
+    private Long userId;
```
JPA 엔티티 매핑 방법에 대해서 더 공부하시는 것을 추천드릴게요.

ksundong commented on BE/src/main/java/com/codesquad/issuetracker/domain/comment/request/CommentRequest.java:
```diff
@@ -0,0 +1,32 @@
+    private String content;
+
+    @JsonProperty("created_at")
+    @JsonFormat(pattern = "YYYY-MM-dd HH:mm:ss")
+    private LocalDateTime createdAt;
```
과연 생성 시간을 클라이언트에서 제공해주는 것이 맞을까요?

ksundong commented on BE/src/main/java/com/codesquad/issuetracker/domain/comment/response/CommentResponse.java:
```diff
@@ -0,0 +1,35 @@
+
+    private UserResponse author;
+
+    public static CommentResponse create(Comment comment) {
```
create보다는 더 적절한 이름을 생각해봅시다.
이펙티브 자바식으로는 `from` 을 많이 사용하긴해요.

ksundong commented on BE/src/main/java/com/codesquad/issuetracker/domain/issue/Issue.java:
```diff
@@ -0,0 +1,74 @@
+
+    private String title;
+    private String content;
+    private boolean status;
```
무슨 status인가요?

ksundong commented on BE/src/main/java/com/codesquad/issuetracker/domain/issue/IssueLabelRepository.java:
```diff
@@ -0,0 +1,20 @@
+import java.util.List;
+
+public interface IssueLabelRepository extends CrudRepository<IssueLabel, Long> {
+    @Query("SELECT il.label.id FROM IssueLabel il WHERE il.issue.id = :issueId")
+    List<Long> findIssueLabelsLabelIdByIssueId(@Param("issueId") Long issueId);
+
+    @Transactional
+    @Modifying
+    @Query("DELETE FROM IssueLabel il WHERE il.issue.id = :issueId AND il.label.id = :labelId")
+    void deleteIssueLabelByIssueIdAndLabelId(@Param("issueId") Long issueId, @Param("labelId") Long labelId);
```
QueryDSL을 사용해보는 것은 어땠을까요?

ksundong commented on BE/src/main/java/com/codesquad/issuetracker/domain/issue/IssueRepositorySupport.java:
```diff
@@ -0,0 +1,102 @@
+
+    }
+
+    public Predicate getPredicate(IssueFilter filter) {
+        return new PredicateBuilder(issue.isNotNull())
+                .containsAnd(issue.title, filter.getTitle())
+                .isBooleanAnd(issue.status, filter.getStatus())
+                .isLongEqualAnd(issue.user.id, filter.getAuthor())
+                .build();
+    }
```
메서드 명이 명확하지 않아 코드를 읽는데 어려움이 있네요.
그리고, public 메서드여야 하는지도 의문이 들어요.

ksundong commented on BE/src/main/java/com/codesquad/issuetracker/domain/issue/IssueRepositorySupport.java:
```diff
@@ -0,0 +1,102 @@
+                    .where(assignee.userId.eq(userId));
+        }
+
+        userId = issueFilter.getCommenter();
```
변경하지 않고 각각을 다른 변수에 담으면 가독성이 높아질거에요.

ksundong commented on BE/src/main/java/com/codesquad/issuetracker/domain/issue/PredicateBuilder.java:
```diff
@@ -0,0 +1,42 @@
+        this.predicate = predicate;
+    }
+
+    public PredicateBuilder containsAnd(StringPath qValue, String rValue) {
```
변수명이 어디서 유래한건지 모르겠지만 잘 와닿지는 않네요.

ksundong commented on BE/src/main/java/com/codesquad/issuetracker/domain/issue/controller/IssueController.java:
```diff
@@ -0,0 +1,110 @@
+        this.issueService = issueService;
+    }
+
+    @GetMapping("/open")
+    public ApiResponse getOpenedIssues() {
+        return ApiResponse.ok(issueService.getOpenedIssues());
+    }
+
+    @GetMapping("/close")
+    public ApiResponse getClosedIssues() {
+        return ApiResponse.ok(issueService.getClosedIssues());
+    }
```
상태를 이용해서 검색할 때에는 query param을 이용하는 것을 더 추천드립니다.

ksundong commented on BE/src/main/java/com/codesquad/issuetracker/domain/issue/service/IssueService.java:
```diff
@@ -0,0 +1,236 @@
+    private final AssigneeRepository assigneeRepository;
+    private final CommentRepository commentRepository;
+
+    @Autowired
```
Autowired 붙여주신 이유가 있나요?

ksundong commented on BE/src/main/java/com/codesquad/issuetracker/domain/label/request/LabelRequest.java:
```diff
@@ -0,0 +1,19 @@
+    private String color;
+
+    public Label create(){
+        return Label.create(null, title, content, color);
```
null을 인자로 넘겨줘야만 한다면 그냥 생성자를 새로 하나 정의해주세요.

ksundong commented on BE/src/main/resources/application-jwt.properties:
```diff
@@ -0,0 +1,2 @@
+security.jwt.token.secret-key=Vu6WXrg3t9
```
시크릿 키는 노출되지 않는게 좋겠죠.

ksundong reviewed (changes requested):
안녕하세요. Woody, 비비!! 리뷰어 Dion입니다.
고생많으셨습니다. 리뷰가 늦어버렸네요 ㅎㅎ..

전반적으로 조금은 장황한 코드라는 느낌을 받았습니다.
객체지향적인 특성을 살려서 메서드 분리와 클래스 분리를 해봐도 좋을 것 같아요.
코드스쿼드 6개월간 정말 고생 많으셨고, 앞으로도 열심히 노력해서 현업에서 뵐 수 있으면 좋겠네요.

jihye-woo committed:
refactor : queryDSL 플러그인 삭제 #19

> Annotation Processor 설정으로 대체
  - 참고 링크 : http://honeymon.io/tech/2020/07/09/gradle-annotation-processor-with-querydsl.html

jihye-woo committed:
refactor : ACCESS_TOKEN_TYPE 상수 필드 삭제 #19

> 사용되지 않는 ACCESS_TOKEN_TYPE 상수 필드 삭제

jihye-woo committed:
refactor : 토큰 타입 확인 로직 수정 #19

> "Bearer"로 시작하는 문자열을 찾는 대신, "Bearer" 문자열과 완벽하게 일치할 경우에 JWT 토큰을 추출한다.

jihye-woo committed:
refactor : 토큰 타입 정보가 없을 경우 Optional 처리 #19

> AuthorizationExtractor::extract
  - 리턴 타입을 Optional로 변경

jihye-woo committed:
refactor : "user" 문자열 상수처리 #19

> "user"문자열을 USER 상수 처리해서 user http header로 사용

jihye-woo committed:
refactor : BearerAuthInterceptor::getDefaultUser 삭제 #19

> 사용하지 않는 메소드 삭제
  - JWT 토큰 검증을 하지 않을 때 테스트 용도로 사용하던 메소드 삭제

jihye-woo committed:
fix : JwtProvider 생성자 파라미터 오타 수정 #19

> secretKey 파라미터 오타 수정
  - @Value("${security.jwt.token.secret-key}")
  - & -> $

> application-jwt.properties의 secret key 암호화
  - jasypt 사용

jihye-woo committed:
chore : 로깅 레벨 변경 #19

> JwtProvider::createJwt에서 로그 데이터에 남길 필요가 없는 정보는 debug 레벨로 수정

jihye-woo committed:
refactor : LoginController 생성자 주입 대상 변경 #19

> 생성자를 통해 Environment 빈을 주입받는 대신, @Value 어노테이션을 통해 직접 특정 값을 주입받아서 사용

bibi6666667 committed:
refactor : 로그인 과정의 중복 로직 제거 #19

> WEB과 iOS 로그인 로직 상의 중복 코드들을 리팩토링을 통해 제거함.

bibi6666667 committed:
refactor : 타입 안정성을 위한 UserPlatform enum 생성 #19

> 플랫폼의 요청을 enum으로 판별하고, 유효하지 않은 요청은 InvalidPlatformException으로 처리

bibi6666667 committed:
refactor : GitHubUser의 변수 login -> loginId로 수정 #19

> GitHubUser의 변수명을 수정

bibi6666667 committed:
chore : 가독성을 위한 공백 추가 #19

bibi6666667 committed:
chore : 변수명을 CamelCase로 변경 #19

bibi6666667 committed:
chore : 주석 코드 제거 및 불필요한 코드 제거 #19

bibi6666667 committed:
chore : 응답 데이터로 엔티티 대신 DTO클래스로 수정 #19

> User 대신 UserResponse로 응답하도록 변수 타입 변경
> 컨벤션에 맞게 static 메서드의 위치 조정

bibi6666667 committed:
refactor : 필드 값 주입 방식을 생성자 주입으로 변경 #19

bibi6666667 committed:
feat : Label생성자 추가

> id를 인자로 받지 않는 Label 생성자 추가
> 추가한 생성자를 LabelRequest에서 사용

bibi6666667 committed:
refactor : 클래스명 변경 및 메서드 분리

> GitHubLoginService -> GitHubOauthService 로 클래스명 변경
> LoginService 클래스를 추가해 signIn(), signUp() 메서드 분리

bibi6666667 committed:
feat : Jwt인증에 필요한 래퍼 클래스 추가

> Jwt인증에 필요한 GitHubUser, tokenType 정보를 클래스로 분리

bibi6666667 committed:
refactor : CommentResponse의 create() 메서드명을 from()로 수정

bibi6666667 committed:
feat : Issue의 변수명 status를 isOpen으로 수정

bibi6666667 committed:
refactor : IssueRepositorySupport의 변수명 및 접근제어자 수정

> getFilteredIssues()의 userId를 assigneeUserId와 commenterUserId로 분리해 선언

> getPredicate() 의 접근제어자 수준을 private로 높임

bibi6666667 committed:
chore : 코드 재정렬

bibi6666667 committed:
Merge pull request #20 from bibi6666667/be/review-working

[BE] PR 코드리뷰 피드백 반영
